File: Java/atmCardValidation.java
import java.util.Scanner;

public class atmCardValidation {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        System.out.println("Enter a credit card number: ");
        String cardNumber = input.nextLine().trim();

        // Remove any spaces or hyphens
        cardNumber = cardNumber.replaceAll("[\\s-]", "");

        // Validate input
        if (!isNumeric(cardNumber) || cardNumber.length() < 13 || cardNumber.length() > 16) {
            System.out.println("Invalid input: Must be 13–16 digits.");
            input.close();
            return;
        }

        // Determine card type
        String cardType = getCardType(cardNumber);

        // Validate using Luhn algorithm
        boolean isValid = isValidLuhn(cardNumber);

        // Output results
        System.out.println("Card Type: " + cardType);
        System.out.println("Validity Status: " + (isValid ? "Valid" : "Invalid"));

        input.close();
    }

    // Check if string contains only digits
    private static boolean isNumeric(String str) {
        return str.matches("\\d+");
    }

    // Determine card type based on starting digits and length
    private static String getCardType(String cardNumber) {
        if (cardNumber.startsWith("4")) {
            return "Visa";
        } else if (cardNumber.startsWith("5")) {
            return "MasterCard";
        } else if (cardNumber.startsWith("37")) {
            return "American Express";
        } else if (cardNumber.startsWith("6")) {
            return "Discover";
        } else {
            return "Unknown";
        }
    }

    // Validate card number using Luhn algorithm
    private static boolean isValidLuhn(String cardNumber) {
        int[] digits = cardNumber.chars().map(ch -> ch - '0').toArray();
        int sum = 0;

        // Step 1: Double every second digit from right to left
        for (int i = digits.length - 2; i >= 0; i -= 2) {
            int doubled = digits[i] * 2;
            // If result is two digits, sum them
            digits[i] = (doubled > 9) ? (doubled - 9) : doubled;
            sum += digits[i];
        }

        // Step 2: Add digits in odd positions (1-based, right to left)
        for (int i = digits.length - 1; i >= 0; i -= 2) {
            sum += digits[i];
        }

        // Step 3: Check if sum is divisible by 10
        return sum % 10 == 0;
    }
}

File: JavaScript/atmCardValidation.js
const ATM_Card = {

	visa(card_number) {
	if (!card_number || card_number.length !== 16) {
	return "Valid: false, reason: Invalid length";
}
	if (card_number[0] !== 4) {
	return "Valid: false, reason: Invalid starting digit";
}
	if (!this.isValidLuhn(card_number)) {
	return "Valid: false, reason: Invalid checksum";
}
	return "Valid: true, issuer: Visa Card";
},

	masterCard(card_number) {
	if (!card_number || card_number.length !== 16) {
	return "Valid: false, reason: Invalid length";
}
	if (card_number[0] !== 5) {
	return "Valid: false, reason: Invalid starting digit";
}
	if (!this.isValidLuhn(card_number)) {
	return "Valid: false, reason: Invalid checksum";
}
	return "Valid: true, issuer: MasterCard";
},

	discover(card_number) {
	if (!card_number || card_number.length !== 16) {
	return "Valid: false, reason: Invalid length";
}
	if (card_number[0] !== 6) {
	return "Valid: false, reason: Invalid starting digit";
}
	if (!this.isValidLuhn(card_number)) {
	return "Valid: false, reason: Invalid checksum";
}
	return "Valid: true, issuer: Discover Card";
},

	americanExpress(card_number) {
	if (!card_number || card_number.length !== 15) {
	return "Valid: false, reason: Invalid length";
}
	if (card_number[0] !== 3) {
	return "Valid: false, reason: Invalid starting digit";
}
	if (!this.isValidLuhn(card_number)) {
	return "Valid: false, reason: Invalid checksum";
}
	return "Valid: true, issuer: America Express Card";
},

	isValidLuhn(card_number) {
	let sum = 0;
	let isEven = false;
	for (let i = card_number.length - 1; i >= 0; i--) {
	let digit = card_number[i];
	if (isEven) {
	digit *= 2;
	if (digit > 9) digit -= 9;
}
	sum += digit;
	isEven = !isEven;
}
	return sum % 10 === 0;
}
};

const Visa_card = [4, 9, 0, 7, 2, 0, 9, 4, 7, 0, 9, 4, 2, 0, 9, 3];
const Master_card = [5, 9, 0, 7, 2, 0, 9, 4, 7, 0, 9, 4, 2, 0, 9, 3];
const Discover_card = [6, 7, 8, 4, 6, 2, 4, 8, 2, 4, 6, 2, 4, 2, 9, 4];
const American_express_card = [3, 9, 0, 7, 2, 0, 9, 4, 7, 0, 9, 4, 2, 0, 9];

console.log(ATM_Card.visa(Visa_card));
console.log(ATM_Card.masterCard(Master_card));
console.log(ATM_Card.discover(Discover_card));
console.log(ATM_Card.americanExpress(American_express_card));